### PR TITLE
CM_Form_Abstract autoSave param

### PR DESCRIPTION
**Pull request by vogdb**
Via https://github.com/cargomedia/sk/issues/4201

### CODE REVIEW

**Comment by njam** [comment]
> shall we pass `autoSave` through back-end? Like:

I would say yes. Let's call it just "autosave" (lower case "s"), because the word describes a concept on its own.

> Shall `autoSave` exclude submit buttons? Theoretically we can have auto-save for one action and button-save for another. Do we need such complication?

I would not cover this case. But I can imagine that it will somewhat okay-ish by default. But it's not a required use case.

> Should we add `data-formfield-success` only when the form was submitted on auto-save or anytime the form was submitted despite the method? 

This attribute should be added to a field *only when the form was autosaved*. The reasoning is that only then we now the specific field that we responsible for the "success".

**Comment by vogdb** [comment]
So, the final logic is:
- we add `data-formfield-error` to the field anytime the field has an error.
- We add `data-formfield-success` only when the field is autosaved successfully.
- Html form representation has `data-autosave="true"`
- `autosave` property transfers form action from back-end to front-end

@njam please look

**Comment by njam** [comment]
@christopheschwyzer wanna review?

File: layout/default/FormField/Abstract/default.less [hunk 1]
```diff
@@ -1,4 +1,4 @@
-&.hasError {
+&.data[formfield-error=true] {
```
Comment (christopheschwyzer) [inline layout/default/FormField/Abstract/default.less]: Selector should be: `&[data-formfield-error=true]`

File: library/CM/Form/Abstract.js [hunk 2]
```diff
@@ -49,10 +52,22 @@ var CM_Form_Abstract = CM_View_Abstract.extend({
...
-    this.$el.on('submit', function() {
-      handler.$el.find('input[type="submit"], button[type="submit"]').first().click();
-      return false;
-    });
+    if (this.autosave) {
+      this.on('change', function(field) {
+        if (field) {
+          handler.submit(handler.autosave).then(function() {
+            if (field.getValue()) {
+              field.success();
```
Comment (christopheschwyzer) [inline library/CM/Form/Abstract.js]: I think we shouldn't trigger success but rather do a validation first? E.g. when you enter "foo" into an email field it shouldn't be highlighted as success.

**Comment by vogdb** [comment]
> I think we shouldn't trigger success but rather do a validation first? E.g. when you enter "foo" into an email field it shouldn't be highlighted as success.

@christopheschwyzer @njam I want to remind that this works only when the submitted form is valid at any time. This happens cause we submit form with `handleErrors: false`. It means that in case of failure we receive all errors as one big text message. It will be very hacky to extract the exact error message of the target field. So in case if two fields fail during validation, we will show the error message for both of them in the autosaved field. If we use `handleErrors: true` then the form will handle and show errors on its own which is good but this will always make `form.submit` to resolve successfully even in case of validation errors. So, we can't detect 100% whether the autosaved field was submitted successfully.

@christopheschwyzer please look.

**Comment by njam** [comment]
It sounds like it would be better to use `handleErrors: true` to have proper error handling.
How about changing the `submit()` function to *even* reject the promise in that case? We would need to check the usages. But it doesn't look like it would be a big problem.

**Comment by vogdb** [comment]
> It sounds like it would be better to use `handleErrors: true` to have proper error handling...

We need to duplicate a `catch` then in all places where we submit manually from js. We have 15 such places. In all of them we would need
```js
form.submit('action').catch(
  CM_Exception_FormFieldValidation,
  function() {
    if(options.handleErrors) {
      // do nothing cause it is handled
    }
  }
)
```

Does not look so nice :) We also can have alternatives. We can split `submit` on two methods: `submit and handle validation` and `submit only`. Of course the first is based on the second. In those 15 places we will use the first. In autosave we will use the second.
Another solution would be to have a method on Form like `getCurrentErrors` that returns all errors the form has now. So anytime we catch we can retrieve the map with errors.

**Comment by vogdb** [comment]
According to the discussion we shall split `submit` on two methods. One for submit (including triggering of events), another for handling errors of submit.

**Comment by vogdb** [comment]
I am sorry. I've rebased the PR to make a cleaner history. Now it can be easily reviewed a commit by commit and does not contain any back and forth redundancy.
key changes:
 - `trigger('complete')` is removed. CM_Form_Abstract does not trigger such event anymore cause it is not used at all and this event didn't fit into the logic of splitting :(
 - `required` errors that previously didn't trigger `error` event, now trigger it.

@njam FYI. @christopheschwyzer please review.

**Comment by christopheschwyzer** [comment]
I added improved form feedback. 
I tested a bit and it seems fine - although I don't understand 100% what's going on. I understand that _autosave_ only works for form actions without any required field. Can we somehow make this fact obvious?

**Comment by vogdb** [comment]
> I understand that _autosave_ only works for form actions without any required field. Can we somehow make this fact obvious?

Nope. Actually now it can work with any form :) due to the implemented split of `submit` 🎉

**Comment by njam** [comment]
Looks great! Agree to release on Monday though.

**Pull request merged**